fin-advisor-crm/markets/views.py

```diff
@@ -1,47 +1,239 @@
+from django.db.models import Q
+from django.views.decorators.csrf import csrf_exempt
 
 from .models import DailyPublicEquityTrades
 
-from datetime import datetime
+import json
+import datetime
+import requests
+import locale
 
 
-# Create your views here.
-def get_summary_card_info():
+import os
+from langchain_core.prompts import ChatPromptTemplate
+from langchain_ollama.llms import OllamaLLM
+
+
+locale.setlocale(locale.LC_ALL, 'en_US.UTF-8')
+
+
+EQUITY_DB_DATE_COL = "trade_date"
+EQUITY_DB_PRICE_COL = "closing_price"
+
+
+def format_currency(value):
+    if value >= 1_000_000:
+        return "GHS{:,.0f} M".format(value / 1_000_000)
+    elif value >= 1_000:
+        return "GHS{:,.0f} K".format(value / 1_000)
+    else:
+        return "GHS{:,.0f}".format(value)
+
+
+def get_dropdown_data(request):
+    data = DailyPublicEquityTrades.objects.all()  # Retrieve all objects from the model
+    tickers = data.values_list("ticker", flat=True)
+    unique_tickers = list(set(tickers))
+    print("drop down func:", unique_tickers)
+    return JsonResponse({'unique_tickers': unique_tickers})
+
+
+def get_ticker_info(share_code):
+    ticker_url = f'https://dev.kwayisi.org/apis/gse/equities/{share_code}'
+    ticker_pull = requests.get(ticker_url)
+    if ticker_pull.ok:
+        return ticker_pull.json()
+    else:
+        raise ValueError
+
+
+def get_ticker_info_basic(share_code):
+    ticker_url = f"https://dev.kwayisi.org/apis/gse/live/{share_code}"
+    ticker_pull = requests.get(ticker_url)
+    if ticker_pull.ok:
+        return ticker_pull.json()
+    else:
+        raise ValueError
+
+
+def equity_trend_data_pull(
+    selected_ticker, 
+    as_of_date
+    ):
+    earliest_date = DailyPublicEquityTrades.objects.earliest("trade_date")
+    start_date = earliest_date.trade_date.strftime("%Y-%m-%d")
+    print("start date: ", earliest_date)
+    selected_ticker_upper = selected_ticker.upper()
+    if as_of_date:
+        as_of_date = datetime.datetime.strptime(
+            as_of_date, 
+            "%Y-%m-%d"
+        ).date()
+        per_ticker_data = DailyPublicEquityTrades.objects.filter(
+            ticker=selected_ticker_upper, 
+            trade_date__range=[start_date, as_of_date]
+        )
+    else:
+        per_ticker_data = DailyPublicEquityTrades.objects.filter(
+            ticker=selected_ticker_upper
+        )
+    return per_ticker_data     
+
+
+def equity_snap_data_pull(
+    selected_ticker, 
+    as_of_date
+    ):
+    if as_of_date:
+        as_of_date = datetime.datetime.strptime(
+            as_of_date, 
+            "%Y-%m-%d"
+        ).date()
+        last_record = DailyPublicEquityTrades.objects.filter(
+            ticker=selected_ticker, 
+            trade_date=as_of_date
+        ).latest("trade_date")
+    else:
+        last_record = DailyPublicEquityTrades.objects.filter(
+            ticker=selected_ticker
+        ).latest("trade_date")
+    return last_record
```

Comment:
Here it seems you should be calling `equity_trend_data_pull` and then calling `.latest("trade_date")` on the result. No need to rewrite a function that does this.